network/application/MQTT_eclipse_paho/mqtt_noos_support.c

```diff
@@ -152,16 +153,22 @@ int mqtt_noos_read(Network* net, unsigned char* buff, int len, int timeout)
 	uint32_t	sent;
 	int32_t		rc;
 
+	if (!len)
+		return 0;
 	sent = 0;
-	while (timeout--) {
-		rc = socket_recv(net->sock, (void *)buff, (uint32_t)len);
-		if (IS_ERR_VALUE(rc))
-			return rc;
-		sent += rc;
-		if (sent >= len)
-			return sent;
+	do {
+ 		rc = socket_recv(net->sock, (void *)(buff + sent),
+				(uint32_t)(len - sent));
+ 		if (rc != EAGAIN) { //If data available or error
+ 			if (IS_ERR_VALUE(rc))
+ 				return rc;
+ 			sent += rc;
+ 			if (sent >= len)
+ 				return sent;
+ 		}
+
 		mdelay(1);
-	}
+	} while (--timeout);
```

Comment:
I think you should decrement your timeout only in case of error.

With the current approach:
Assume you have a timeout of 5 and first 3 loops you get an error. 
Also assume your message is split across 3 messages.
So you get 3 times error + 2 valid messages which hits the timeout. 
If you had done 1 more loop you'd have gotten the full message.

Reply:
If the functions returns 10 bytes but you requested 100, if you call the function again will not return more, that's way delay should be on success too.